mls-rs/src/group/message_processor.rs

```diff
@@ -102,73 +81,48 @@ pub(crate) fn path_update_required(proposals: &ProposalBundle) -> bool {
+)]
 #[derive(Clone, Debug, PartialEq)]
-pub struct StateUpdate {
-    pub(crate) roster_update: RosterUpdate,
-    #[cfg(feature = "psk")]
-    pub(crate) added_psks: Vec<ExternalPskId>,
-    pub(crate) pending_reinit: Option<CipherSuite>,
-    pub(crate) active: bool,
-    pub(crate) epoch: u64,
-    #[cfg(feature = "custom_proposal")]
-    pub(crate) custom_proposals: Vec<ProposalInfo<CustomProposal>>,
-    #[cfg(feature = "by_ref_proposal")]
-    pub(crate) unused_proposals: Vec<crate::mls_rules::ProposalInfo<Proposal>>,
-}
-
-#[cfg(not(feature = "state_update"))]
 #[non_exhaustive]
-#[derive(Clone, Debug, PartialEq)]
-pub struct StateUpdate {}
-
-#[cfg(feature = "state_update")]
-impl StateUpdate {
-    /// Changes to the roster as a result of proposals.
-    pub fn roster_update(&self) -> &RosterUpdate {
-        &self.roster_update
-    }
-
-    #[cfg(feature = "psk")]
-    /// Pre-shared keys that have been added to the group.
-    pub fn added_psks(&self) -> &[ExternalPskId] {
-        &self.added_psks
-    }
+pub struct NewEpoch {
+    pub epoch: u64,
+    pub prior_state: GroupState,
+    pub applied_proposals: Vec<ProposalInfo<Proposal>>,
+    pub unused_proposals: Vec<ProposalInfo<Proposal>>,
+}
 
-    /// Flag to indicate if the group is now pending reinitialization due to
-    /// receiving a [`ReInit`](crate::group::proposal::Proposal::ReInit)
-    /// proposal.
-    pub fn is_pending_reinit(&self) -> bool {
-        self.pending_reinit.is_some()
+#[cfg(all(feature = "ffi", not(test)))]
+#[safer_ffi_gen::safer_ffi_gen]
+impl NewEpoch {
+    pub fn epoch(&self) -> u64 {
+        self.epoch
     }
 
-    /// Flag to indicate the group is still active. This will be false if the
-    /// member processing the commit has been removed from the group.
-    pub fn is_active(&self) -> bool {
-        self.active
+    pub fn prior_state(&self) -> &GroupState {
+        &self.prior_state
     }
 
-    /// The new epoch of the group state.
-    pub fn new_epoch(&self) -> u64 {
-        self.epoch
+    pub fn applied_proposals(&self) -> &[ProposalInfo<Proposal>] {
+        &self.applied_proposals
     }
 
-    /// Custom proposals that were committed to.
-    #[cfg(feature = "custom_proposal")]
-    pub fn custom_proposals(&self) -> &[ProposalInfo<CustomProposal>] {
-        &self.custom_proposals
-    }
-
-    /// Proposals that were received in the prior epoch but not committed to.
-    #[cfg(feature = "by_ref_proposal")]
-    pub fn unused_proposals(&self) -> &[crate::mls_rules::ProposalInfo<Proposal>] {
+    pub fn unused_proposals(&self) -> &[ProposalInfo<Proposal>] {
         &self.unused_proposals
     }
+}
 
-    pub fn pending_reinit_ciphersuite(&self) -> Option<CipherSuite> {
-        self.pending_reinit
-    }
+#[cfg_attr(
+    all(feature = "ffi", not(test)),
+    safer_ffi_gen::ffi_type(clone, opaque)
+)]
+#[derive(Clone, Debug, PartialEq)]
+pub enum CommitEffect {
+    NewEpoch(Box<NewEpoch>),
```

Comment:
Out of curiosity, did clippy suggest to box it? Just asking because it makes pattern matching more difficult.

Reply:
Yeah, you have to Box this because the difference in size is very large. Something like 600 bytes

Reply:
Got it. Thank you.